CMakeLists.txt

```diff
@@ -1,35 +1,28 @@
-cmake_minimum_required(VERSION 3.5.1)
+cmake_minimum_required(VERSION 3.7)
 
 find_program(CCACHE_PROGRAM ccache)
 if(CCACHE_PROGRAM)
   set_property(GLOBAL PROPERTY RULE_LAUNCH_COMPILE "${CCACHE_PROGRAM}")
+  set_property(GLOBAL PROPERTY RULE_LAUNCH_LINK    "${CCACHE_PROGRAM}")
   message(STATUS "ccache enabled (${CCACHE_PROGRAM})")
 endif()
 
-PROJECT(iroha C CXX)
-
-SET(CMAKE_POSITION_INDEPENDENT_CODE TRUE)
-SET(CMAKE_CXX_FLAGS "-std=c++1y -Wall")
-SET(CMAKE_CXX_FLAGS_RELEASE "-O3")
-SET(CMAKE_CXX_FLAGS_DEBUG   "-g -Wextra -Wno-unused-parameter -Wno-deprecated-declarations -O0")
-SET(CMAKE_CXX_OUTPUT_EXTENSION_REPLACE 1)
-SET(CMAKE_INSTALL_RPATH "../lib")
-
-if(COVERAGE)
-  # remove -g flag to reduce binary size
-  list(REMOVE_ITEM CMAKE_CXX_FLAGS -g)
-  set(CMAKE_EXPORT_COMPILE_COMMANDS "ON")
-  set(CMAKE_CXX_FLAGS "--coverage ${CMAKE_CXX_FLAGS}")
-  set(CMAKE_C_FLAGS "--coverage ${CMAKE_C_FLAGS}")
-  set(CMAKE_SHARED_LINKER_FLAGS "--coverage ${CMAKE_SHARED_LINKER_FLAGS}")
-  set(CMAKE_EXE_LINKER_FLAGS "--coverage ${CMAKE_EXE_LINKER_FLAGS}")
-
-  set(REPORT_DIR ${CMAKE_BINARY_DIR}/reports)
-  file(MAKE_DIRECTORY ${REPORT_DIR})
-  include(cmake/analysis.cmake)
-endif()
+set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake/Modules)
+
+#########################
+## configuration types ##
+#########################
+set(CMAKE_CONFIGURATION_TYPES "Debug;Release;Coverage" CACHE STRING "" FORCE)
```

Comment:
It is not required to modify CMake targets, because coverage only modifies flag set - it can be used both in Debug and Release.

Reply:
It is convenient to group flags as configuration types. The main reason is that you can select specific configuration at cmake time 

```
cmake .. -DCMAKE_BUILD_TYPE=Coverage
```